go/ethadapter/blockprovider.go

```diff
@@ -0,0 +1,272 @@
+type LiveBlocksMonitor struct {
+	ctx       context.Context
+	logger    gethlog.Logger
+	ethClient EthClient
+
+	latest *types.Header
+
+	awaitChan chan *types.Header // recreated when user starts awaiting, receives true when new head
+}
+
+func (l *LiveBlocksMonitor) Start() {
+	blkHeadChan, blkSubs := l.ethClient.BlockListener()
+	for {
+		select {
+		case <-l.ctx.Done():
+			return
+
+		case blkHead := <-blkHeadChan:
+			l.logger.Trace("received new L1 head", "height", blkHead.Number, "hash", blkHead.Hash())
+			l.latest = blkHead
+			if l.awaitChan != nil {
+				l.awaitChan <- blkHead // notify
+			}
+
+		case err := <-blkSubs.Err():
+			l.logger.Error("L1 block monitoring error", log.ErrKey, err)
+			l.logger.Info("Restarting L1 block Monitoring...")
+			// this disconnect could result in a gap in the LiveBlocksMonitor queue, but the block provider is responsible
+			// for sending a coherent ordering of blocks and will look up blocks if parent not sent
+			blkHeadChan, blkSubs = l.ethClient.BlockListener()
+		}
+	}
+}
+
+// AwaitNewBlock takes a hash, it will block until it can return a head block with a different hash or error
+// (note: this can currently only be used by one caller at a time - not an issue for current usage)
+func (l *LiveBlocksMonitor) AwaitNewBlock(ctx context.Context, afterBlkHash gethcommon.Hash) (*types.Header, error) {
+	awaitChan := l.registerAwait()
+	streamTimeout := 1 * time.Minute
+	if l.latest == nil {
+		// note: the in-mem client does not support streaming from ethClient.BlockListener() currently, this allows it to still function
+		l.logger.Debug("latest is nil, blockprovider will timeout after 1sec without a block as client streaming may not be working")
+		streamTimeout = 1 * time.Second
+	} else {
+		l.logger.Trace("latest=%s, afterBl=%s\n", l.latest.Hash(), afterBlkHash)
+	}
+	if l.latest != nil && l.latest.Hash() != afterBlkHash {
+		// return immediately if caller is not at latest head
+		return l.latest, nil
+	}
+
+	// otherwise wait for a block to show up
+	for {
+		select {
+		case blkHead := <-awaitChan:
+			return blkHead, nil
+		case <-ctx.Done():
+			return nil, errors.New("context expired before block found")
+		case <-time.After(streamTimeout):
+			if l.latest == nil {
+				// not seen any streamed blocks, see if we can get head from client
+				// note: in-mem sim client doesn't support block streaming, it will always use this path
+				block := l.ethClient.FetchHeadBlock()
+				if block.Hash() != afterBlkHash {
+					return block.Header(), nil
+				}
+				continue // not seen any streamed block and head isn't new, wait for the next timeout.
+			}
+			return nil, fmt.Errorf("no live block received after timeout=%s", streamTimeout)
+		}
+	}
+}
+
+func (l *LiveBlocksMonitor) registerAwait() <-chan *types.Header {
+	l.awaitChan = make(chan *types.Header)
+	return l.awaitChan
+}
+
+func (e *EthBlockProvider) latestCanonAncestor(blkHash gethcommon.Hash) (*types.Block, error) {
```

Comment:
this could be a simple function that receives the ethclient and the block hash as parameters.
It's a bit confusing to see it here defined on the `LiveBlocksMonitor` after the `LiveBlocksMonitor` was declared

Reply:
Moved it up to be with the other BlockProvider methods, not sure how it ended up down there.